src/wallet/scriptpubkeyman.cpp

```diff
@@ -732,12 +787,27 @@ bool LegacyScriptPubKeyMan::AddWatchOnly(const CScript& dest, int64_t nCreateTim
 void LegacyScriptPubKeyMan::SetHDChain(const CHDChain& chain, bool memonly)
 {
     LOCK(cs_wallet);
-    if (!memonly && !WalletBatch(m_storage.GetDatabase()).WriteHDChain(chain))
-        throw std::runtime_error(std::string(__func__) + ": writing chain failed");
+    if (!memonly) {
+        if (!WalletBatch(m_storage.GetDatabase()).WriteHDChain(chain)) {
+            throw std::runtime_error(std::string(__func__) + ": writing chain failed");
+        }
+        if (!hdChain.seed_id.IsNull()) {
+            AddInactiveHDChain(hdChain, false);
+        }
+    }
 
     hdChain = chain;
 }
 
+void LegacyScriptPubKeyMan::AddInactiveHDChain(const CHDChain& chain, bool memonly)
+{
+    LOCK(cs_wallet);
+    if (!memonly && !WalletBatch(m_storage.GetDatabase()).WriteInactiveHDChain(chain))
+        throw std::runtime_error(std::string(__func__) + ": writing inactive chain failed");
+
+    m_inactive_hd_chains[chain.seed_id] = chain;
```

Comment:
should we check if `chain.seed_id` is null here?

Reply:
It should never be null. I've added an assert to check that.